File: src/data/blocks.ts
/** ROUNDrBR.CT (72 B/block: 6x6 LE16 tile words), .COL (18 B/block: 1 bit per 8x8 cell, 12x12),
 *  .DIR (36 B/block: 1 byte per tile), .LEV (1 B/block). Port of tools/mm/blocks.py; semantics in
 *  re/notes/80-ai-collision.md. */
export interface Blocks {
  count: number;
  tiles: Uint16Array;   // count * 36 tile words (bit 15 = priority tile drawn over sprites)
  col: Uint8Array;      // count * 18
  dir: Uint8Array;      // count * 36 (padded/truncated to 36 per block like the game's fixed stride)
  lev: Uint8Array;      // count (missing bytes read as 0)
}

export function decodeBlocks(ct: Uint8Array, col: Uint8Array, dir: Uint8Array, lev: Uint8Array): Blocks {
  if (ct.length % 72) throw new Error('CT size not a multiple of 72');
  const n = ct.length / 72;
  if (col.length !== n * 18) throw new Error(`COL size ${col.length} != ${n}*18`);
  const tiles = new Uint16Array(n * 36);
  for (let i = 0; i < n * 36; i++) tiles[i] = ct[i * 2]! | (ct[i * 2 + 1]! << 8);
  const dirOut = new Uint8Array(n * 36); dirOut.set(dir.subarray(0, Math.min(dir.length, n * 36)));
  const levOut = new Uint8Array(n); levOut.set(lev.subarray(0, Math.min(lev.length, n)));
  return { count: n, tiles, col: new Uint8Array(col), dir: dirOut, lev: levOut };
}

/** 192x192 tile words as the game builds them (fn 447e). Cells whose block index is out of range are left 0
 *  (the game would read stale memory past the CT there). */
export function expandMap(blocks: Blocks, mapBlocks: Uint8Array): Uint16Array {
  const out = new Uint16Array(192 * 192);
  for (let i = 0; i < 1024; i++) {
    const b = mapBlocks[i]! & 0x3F;
    if (b >= blocks.count) continue;
    const bx = i % 32, by = (i / 32) | 0;
    for (let r = 0; r < 6; r++)
      for (let c = 0; c < 6; c++) out[(by * 6 + r) * 192 + bx * 6 + c] = blocks.tiles[b * 36 + r * 6 + c]!;
  }
  return out;
}

File: tools/mm/blocks.py
"""ROUNDrBR.CT (72 B/block: 6x6 LE16 tile words), .COL (18 B/block), .DIR (36 B/block), .LEV (1 B/block)."""
from dataclasses import dataclass
import struct

@dataclass
class Blocks:
    tiles: list   # per block: 36 ints (tile word, bit 15 = priority)
    col: list     # per block: 18 raw bytes
    dir: list     # per block: 36 raw bytes
    lev: list     # per block: 1 byte (may be shorter/longer in the files)

def decode_blocks(ct: bytes, col: bytes, dir_: bytes, lev: bytes) -> Blocks:
    if len(ct) % 72:
        raise ValueError("CT size not a multiple of 72")
    n = len(ct) // 72
    if len(col) != n * 18:
        raise ValueError(f"COL size {len(col)} != {n}*18")
    tiles = [list(struct.unpack_from('<36H', ct, i * 72)) for i in range(n)]
    colb = [list(col[i * 18:(i + 1) * 18]) for i in range(n)]
    dirb = [list(dir_[i * 36:(i + 1) * 36]) for i in range(n)]   # ROUND5 short / ROUND8 long: pad/ignore
    dirb = [d + [0] * (36 - len(d)) for d in dirb]
    levb = list(lev[:n]) + [0] * max(0, n - len(lev))
    return Blocks(tiles, colb, dirb, levb)

def expand_map(blocks: Blocks, map_blocks: list):
    """192x192 tile words as the game builds them (fn 447e)."""
    out = [[0] * 192 for _ in range(192)]
    nblk = len(blocks.tiles)
    for i, b in enumerate(map_blocks):
        if (b & 0x3F) >= nblk:
            continue        # the game would read stale memory past the CT here (ROUND82 has such cells)
        t = blocks.tiles[b & 0x3F]
        bx, by = i % 32, i // 32
        for r in range(6):
            out[by * 6 + r][bx * 6:bx * 6 + 6] = t[r * 6:r * 6 + 6]
    return out
